packages/@aws-cdk/cloud-assembly-schema/README.md

```diff
@@ -0,0 +1,62 @@
+## Cloud Assembly Schema
+<!--BEGIN STABILITY BANNER-->
+
+---
+
+![Stability: Stable](https://img.shields.io/badge/stability-Stable-success.svg?style=for-the-badge)
+
+
+---
+<!--END STABILITY BANNER-->
+
+This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) project.
+
+## Cloud Assembly
+
+The *Cloud Assembly* is the artifact of the AWS CDK class library. It is produced as part of the
+[`cdk synth`](https://github.com/aws/aws-cdk/tree/master/packages/aws-cdk#cdk-synthesize)
+command, or the [`app.synth()`](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/core/lib/app.ts#L135) method invocation.
+
+Its essentially a set of files and directories, one of which is the `manifest.json` file. It defines the set of instructions that are
+needed in order to deploy the assembly directory.
+
+> For example, when `cdk deploy` is executed, the CLI reads this file and performs its instructions:
+> - Build container images.
+> - Upload assets.
+> - Deploy CloudFormation templates.
+
+Therefore, the assembly is how the CDK class library and CDK CLI (or any other consumer) should communicate. To ensure compatibility
+between the assembly and its consumers, we treat the manifest file as a well defined, versioned schema.
+
+## Schema
+
+This module contains the typescript structs that comprise the `manifest.json` file, as well as the
+generated [*json-schema*](./schema/cloud-assembly.schema.json).
+
+## Versioning
+
+The schema version is specified in the [`cloud-assembly.metadata.json`](./schema/cloud-assembly.schema.json) file, under the `version` property.
+It follows semantic versioning, but with a small twist.
+
+When we add instructions to the assembly, they are reflected in the manifest file and the *json-schema* accordingly.
+Every such instruction, is crucial for ensuring the correct deployment behavior. This means that to properly deploy a cloud assembly,
+consumers must be aware of every such instruction modification.
+
+For this reason, every change to the schema, even though it might not strictly break validation of the *json-schema* format,
+is considered `major` version bump.
+
+## How to consume
+
+If you'd like to consume the schema in order to do validations on `manifest.json` files, simply download it from this repo and run it against
```

Comment:
> 'download it' - how?

I can add an example `curl` command. 

> Instead of downloading it from the repo, perhaps it should be made available in npm package?

The file is available in the npm package of `cloud-assembly-schema` - but how would you access it specifically? In general, specs should actually be hosted in some domain that we own, i.e 'https://aws-cdk.io/cloud-assembly-schema/2.0.0`...
But its not entirely uncommon to simply fetch schemas from `raw.github`.  

We might setup this domain at some point.

Reply:
> but how would you access it specifically?

npm install ?

Wouldn't that be more accurate, in that, it would contain the latest *released* schema and not unreleased schema on master?

I am assuming here, a consumer wants to consume it programmatically for their own application. In this case, they are going to take a dependency on the cloud-assembly-schema package and that's their consumption model.

This is what we should be saying here, no?

Reply:
Not necessarily. Its meant as a consumable independent json **file**,  that can be used from any language. That was the general use case i had in mind when i wrote "download". Similarly to how we download the CloudFormation spec.

Reply:
> Similarly to how we download the CloudFormation spec

We don't download the spec from their source code, instead from their website which is the latest published version of the spec. The equivalent for that is to obtain from this from npm repo or github repo, is it not?

Further, we download the cfn spec from the website because CF publishes it there. If CF were to publish this into a package into an artifactory like npm, we would take a dependency on this package and consume it that way, no?

Reply:
> we would take a dependency on this package

If we happen to write in a language that they publish a package for, sure.

> We don't download the spec from their source code, instead from their website which always contains the latest published version of the spec

This is exactly what we should be doing at some point. 
Until then, the only way to download this specific json schema file (that is language agnostic) is by pointing to github as "our website".

> The equivalent for that is to obtain from this from npm repo or github repo.

Well, you can't really download a single file from NPM, so yeah, github repo it is.